src/daemon.rs

```diff
@@ -533,8 +597,23 @@ impl Daemon {
         verbose: bool,
     ) -> Result<Value> {
         let mut args = json!([txhash.to_hex(), verbose]);
-        if let Some(blockhash) = blockhash {
-            args.as_array_mut().unwrap().push(json!(blockhash.to_hex()));
+        let mut conn = self.conn.lock().unwrap();
+        match (conn.txindex, blockhash) {
+            (TxIndexState::Synced, _) => {}
+            (_, None) => {}
+            (TxIndexState::Unavailable, Some(blockhash)) => {
+                args.as_array_mut().unwrap().push(json!(blockhash.to_hex()));
+            }
+            (TxIndexState::Unknown, Some(blockhash)) => {
+                conn.update_txindex_state()?; 
+                args.as_array_mut().unwrap().push(json!(blockhash.to_hex())); 
+            }
+            (TxIndexState::Syncing{last_check}, Some(blockhash)) => {
+                if last_check.elapsed() > Duration::from_secs(3600) {
```

Comment:
1h seems quite pessimistic to me. AFAIR syncing is not terribly slow so I'd prefer trading a little bit of performance during sync for better performance after it. 10 mins maybe?

Reply:
I think I was worried about the cost of calling `getindexinfo` for each transaction (the reason why we introduced TxIndexState at the first place). But 10 min seems fair too, this call is quite fast after all :)